misk/src/main/kotlin/misk/digester/MergingDigest.kt

```diff
@@ -0,0 +1,351 @@
+      builder.mean +=
+          (next.mean - mainCentroids[mainCentroids.size-1].mean) * next.weight / mainCentroids[mainCentroids.size-1].weight
+
+      // we did not create a new centroid, so the trailing index of the previous
+      // centroid remains
+      return beforeIndex
+    }
+  }
+
+  /**
+   * given a quantile, estimate the index of the centroid that contains it using
+   * the given compression
+   */
+  fun indexEstimate(quantile: Double) : Double {
+    return compression * ((asin(2*quantile-1) / PI) + 0.5).toFloat()
+  }
+
+  /**
+   * Returns a value such that the fraction of values in td below that value is
+   * approximately equal to quantile. Returns NaN if the digest is empty.
+   */
+  fun quantile(quantile: Double): Double {
+    if (quantile < 0 || quantile > 1) {
+      throw IndexOutOfBoundsException("quantile out of bounds")
+    }
+    mergeAllTemps()
+
+    // add up the weights of centroids in ascending order until we reach a
+    // centroid that pushes us over the quantile
+    var q = quantile * mainWeight
+    var weightSoFar = 0.0
+    var lowerBound = min
+
+    val iterator = mainCentroids.iterator()
+
+    for((i, c) in iterator.withIndex()) {
+      var upperBound = centroidUpperBound(i)
+      if (q <= weightSoFar+c.weight) {
+        // the target quantile is somewhere inside this centroid
+        // we compute how much of this centroid's weight falls into the quantile
+        var proportion = (q - weightSoFar) / c.weight
+        // and interpolate what value that corresponds to inside a uniform
+        // distribution
+        return lowerBound + (proportion * (upperBound - lowerBound))
+      }
+
+      // the quantile is above this centroid, so sum the weight and carry on
+      weightSoFar += c.weight
+      lowerBound = upperBound
+    }
+
+    // should never be reached unless empty, since the final comparison is
+    // q <= td.mainWeight
+    return Double.NaN
+  }
+
+  /**
+   *  we assume each centroid contains a uniform distribution of values
+   * the lower bound of the distribution is the midpoint between this centroid and
+   * the previous one (or the minimum, if this is the lowest centroid)
+   * similarly, the upper bound is the midpoint between this centroid and the
+   * next one (or the maximum, if this is the greatest centroid)
+   * this function returns the position of the upper bound (the lower bound is
+   * equal to the upper bound of the previous centroid)
+   * this assumption is justified empirically in dunning's paper
+   * TODO: does this assumption actually apply to our implementation?
+   */
+  fun centroidUpperBound(i: Int): Double {
+    if (i != mainCentroids.size-1) {
+      return (mainCentroids[i+1].mean + mainCentroids[i].mean) / 2
+    } else {
+      return max
+    }
+  }
+
+  /**
+   * merge another digest into this one. Neither td nor other can be shared
+   * concurrently during the execution of this method.
+   */
+  fun merge(other: MergingDigest) {
```

Comment:
This should either return a new `MergingDigest` which is the result of the merge (my preference, so that digests can be made immutable) or should indicate in the method name which digest is being modified (e.g. `mergeFrom` if the receiver is being modified, or `mergeInto` if the parameter is being modified)

Reply:
We can keep this mutable since `merge` is an internal method; I left a separate comment about making a bunch of methods private.

Since `MergingDigests` will have many values added to them, I think we should keep them mutable like other metrics-oriented collector types.